nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteProvenanceReportingTask.java

```diff
@@ -0,0 +1,354 @@
+                }
+
+                final Map<String, String> attributes = new HashMap<>();
+                final String transactionId = UUID.randomUUID().toString();
+                attributes.put("reporting.task.transaction.id", transactionId);
+
+                final byte[] data = jsonArray.toString().getBytes(StandardCharsets.UTF_8);
+                transaction.send(data, attributes);
+                transaction.confirm();
+                transaction.complete();
+
+                final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start);
+                getLogger().info("Successfully sent {} Provenance Events to destination in {} ms; Transaction ID = {}; First Event ID = {}",
+                        new Object[]{events.size(), transferMillis, transactionId, events.get(0).getEventId()});
+            } catch (final IOException e) {
+                throw new ProcessException("Failed to send Provenance Events to destination due to IOException:" + e.getMessage(), e);
+            }
+
+            // Store the id of the last event so we know where we left off
+            final ProvenanceEventRecord lastEvent = events.get(events.size() - 1);
+            final String lastEventId = String.valueOf(lastEvent.getEventId());
+            try {
+                StateManager stateManager = context.getStateManager();
+                Map<String, String> newMapOfState = new HashMap<>();
+                newMapOfState.put(LAST_EVENT_ID_KEY, lastEventId);
+                stateManager.setState(newMapOfState, Scope.LOCAL);
+            } catch (final IOException ioe) {
+                getLogger().error("Failed to update state to {} due to {}; this could result in events being re-sent after a restart. The message of {} was: {}",
+                        new Object[]{lastEventId, ioe, ioe, ioe.getMessage()}, ioe);
+            }
+
+            firstEventId = lastEvent.getEventId() + 1;
+
+            // Retrieve the next batch
+            try {
+                events = context.getEventAccess().getProvenanceEvents(firstEventId, context.getProperty(BATCH_SIZE).asInteger());
+            } catch (final IOException ioe) {
+                getLogger().error("Failed to retrieve Provenance Events from repository due to: " + ioe.getMessage(), ioe);
+                return;
+            }
+        }
+
+    }
+
+    static JsonObject serialize(final JsonBuilderFactory factory, final JsonObjectBuilder builder, final ProvenanceEventRecord event,
+        final String componentName, final String hostname, final URL nifiUrl, final String applicationName, final String platform) {
+        addField(builder, "eventId", UUID.randomUUID().toString());
+        addField(builder, "eventOrdinal", event.getEventId());
+        addField(builder, "eventType", event.getEventType().name());
+        addField(builder, "timestampMillis", event.getEventTime());
+
+        final DateFormat df = new SimpleDateFormat(TIMESTAMP_FORMAT);
```

Comment:
I believe SimpleDateFormat is fairly expensive to initialize. Would recommend we pull this out of this function and pass it in as an argument, so that we can define it in the onTrigger method.